ext/standard/streamsfuncs.c

```diff
@@ -616,93 +616,132 @@ PHP_FUNCTION(stream_get_wrappers)
+    (set)->size = (num_fds + 7) / 8; \
+    (set)->fds_bits = (char *)ecalloc((set)->size, sizeof(char)); \
+} while (0)
+#define FD_BIGSET_SET(fd, set) ((set)->fds_bits[(fd) / 8] |= (1 << ((fd) % 8)))
+#define FD_BIGSET_ISSET(fd, set) ((set)->fds_bits[(fd) / 8] & (1 << ((fd) % 8)))
+#define FD_BIGSET_CLR(fd, set) ((set)->fds_bits[(fd) / 8] &= ~(1 << ((fd) % 8)))
+#define FD_BIGSET_FREE(set) do { \
+    if ((set)->fds_bits) { \
+        efree((set)->fds_bits); \
+        (set)->fds_bits = NULL; \
+    } \
+    (set)->size = 0; \
+} while (0)
+#else
+#define FD_BIGSET_ZERO(set, num_fds) do { \
+    (set)->size = (num_fds + 7) / 8; \
+    (set)->__fds_bits = (char *)ecalloc((set)->size, sizeof(char)); \
+} while (0)
+#define FD_BIGSET_SET(fd, set) ((set)->__fds_bits[(fd) / 8] |= (1 << ((fd) % 8)))
+#define FD_BIGSET_ISSET(fd, set) ((set)->__fds_bits[(fd) / 8] & (1 << ((fd) % 8)))
+#define FD_BIGSET_CLR(fd, set) ((set)->__fds_bits[(fd) / 8] &= ~(1 << ((fd) % 8)))
+#define FD_BIGSET_FREE(set) do { \
+    if ((set)->__fds_bits) { \
+        efree((set)->__fds_bits); \
+        (set)->__fds_bits = NULL; \
+    } \
+    (set)->size = 0; \
+} while (0)
+#endif
 
-	if (Z_TYPE_P(stream_array) != IS_ARRAY) {
-		return 0;
-	}
+int stream_array_to_fd_bigset(zval *stream_array, fd_bigset *set, php_socket_t *max_fd, long max_fds) {
+    zval *elem;
+    php_stream *stream;
+    int cnt = 0;
+
+    if (Z_TYPE_P(stream_array) != IS_ARRAY) {
+        return 0;
+    }
 
-	ZEND_HASH_FOREACH_VAL(Z_ARRVAL_P(stream_array), elem) {
+    ZEND_HASH_FOREACH_VAL(Z_ARRVAL_P(stream_array), elem) {
 		/* Temporary int fd is needed for the STREAM data type on windows, passing this_fd directly to php_stream_cast()
 			would eventually bring a wrong result on x64. php_stream_cast() casts to int internally, and this will leave
 			the higher bits of a SOCKET variable uninitialized on systems with little endian. */
 		php_socket_t this_fd;
 
-		ZVAL_DEREF(elem);
-		php_stream_from_zval_no_verify(stream, elem);
-		if (stream == NULL) {
-			continue;
-		}
+        ZVAL_DEREF(elem);
+        php_stream_from_zval_no_verify(stream, elem);
+        if (stream == NULL) {
+            continue;
+        }
 		/* get the fd.
 		 * NB: Most other code will NOT use the PHP_STREAM_CAST_INTERNAL flag
 		 * when casting.  It is only used here so that the buffered data warning
 		 * is not displayed.
 		 * */
 		if (SUCCESS == php_stream_cast(stream, PHP_STREAM_AS_FD_FOR_SELECT | PHP_STREAM_CAST_INTERNAL, (void*)&this_fd, 1) && this_fd != -1) {
-
-			PHP_SAFE_FD_SET(this_fd, fds);
-
-			if (this_fd > *max_fd) {
-				*max_fd = this_fd;
-			}
-			cnt++;
-		}
-	} ZEND_HASH_FOREACH_END();
-	return cnt ? 1 : 0;
+            if (this_fd < max_fds) {  // Check if the descriptor is within the limit
+                FD_BIGSET_SET(this_fd, set);
+                if (this_fd > *max_fd) {
+                    *max_fd = this_fd;
+                }
+            }
+            cnt++;
```

Comment:
No, unless the test suite is updated, this is the behavior it must have. I did first implement it like you expect but then it will fail tests because it changes what errors are emitted. Perhaps it could be improved for PHP 9 or 8.5?

Reply:
If the tests fails when you add a warning that a stream isn't being selected, then there is probably something wrong with this code.

Reply:
I don't feel I can change the semantics of how the function works, which would require changes to the test suite. The code as it stands now provides identical errors as it used to before my PR. The php_stream_cast API allows me to cast a stream resource to a file descriptor id, and it has an argument where I can declare that I don't want it to emit errors. The streamWrapper does not respect this argument, so I can't avoid having it emit an additional warning when I try to search for the largest file descriptor ID. The new version instead allocates enough space to select() any file descriptor ID that is possible within the constraints of the ulimit. Apparently there are situations where the sysconf() function is not available, so I have to look into that.

Reply:
Ah, I see. The previous code tucks away the if-statement in the `PHP_SAFE_FD_SET` macro, this just extracts it. Makes sense now. So, there isn't a big behavior change. However, there still is a subtle one. I think to keep the previous (possibly incorrect?) behavior:

```suggestion
            if (this_fd < max_fds) {  // Check if the descriptor is within the limit
                FD_BIGSET_SET(this_fd, set);
            }
            if (this_fd > *max_fd) {
                *max_fd = this_fd;
            }
            cnt++;
```